README.md
[![build](https://github.com/nexron171/JustBottomSheet/actions/workflows/tests.yml/badge.svg)](https://codecov.io/gh/nexron171/JustBottomSheet)
[![codecov](https://codecov.io/gh/nexron171/JustBottomSheet/branch/master/graph/badge.svg?token=GGJMMAESDU)](https://codecov.io/gh/nexron171/JustBottomSheet)

***Just Bottom Sheet***

<img src="https://devnex-2796b.web.app/just_bottom_sheet/example.gif" height="400"/>

Bottom sheet with properly working nested scroll and swipe to close gesture.

## Features

- Highly customizable
- Close only by drag handler or by drag handler and scroll
- Supports any scrolls inside including slivers
- More features soon

## Usage

Just call function:

```dart
showJustBottomSheet(...);
```
or you can present through your own route:
```dart
Navigator.of(context).push(YourRoute(builder: (context) {
    return JustBottomSheetPage(...);
}));
```

CHANGELOG.md
## 1.3.0

* Fixed main scroll behavior in case of horizontal nested scroll

## 1.2.0

* Ability to use ImageFilter as background
* Disabled scroll while dragging

## 1.1.0

* Some code reorganization
* Drag zone configuration - you can specify position of drag zone: inside of bottom sheet or outside
* Availablity to specify custom handle bar widget

## 1.0.3

* Changed close on scroll logic

## 1.0.2

* 100% code coverage

## 1.0.1

* Fixed analyzer warning
* Added code coverage and build badges

## 1.0.0

* Initial release.


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:just_bottom_sheet/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


integration_test/main_test.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';
import 'package:just_bottom_sheet/main.dart' as app;

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  group('end-to-end test', () {
    testWidgets('tap on the floating action button to open bottom sheet', (WidgetTester tester) async {
      app.main();
      await tester.pumpAndSettle();
      final fab = await _openBottomSheet(tester);

      final draggableZone = find.byKey(const Key("DraggableZone"));
      expect(fab.hitTestable(), findsNothing, reason: "FAB should be invisible");
      expect(draggableZone, findsOneWidget, reason: "Bottom sheet should be opened");
    });

    testWidgets('fling on draggable zone to close bottom sheet', (WidgetTester tester) async {
      app.main();
      await tester.pumpAndSettle();
      final fab = await _openBottomSheet(tester);

      final draggableZone = find.byKey(const Key("DraggableZone"));
      expect(fab.hitTestable(), findsNothing, reason: "FAB should be invisible");
      expect(draggableZone, findsOneWidget, reason: "Bottom sheet should be opened");

      await tester.fling(draggableZone, const Offset(0, 50), 500);
      await tester.pumpAndSettle();
      expect(draggableZone.hitTestable(), findsOneWidget);

      await tester.fling(draggableZone, const Offset(0, 50), 1500);
      await tester.pumpAndSettle();

      expect(fab.hitTestable(), findsOneWidget, reason: "FAB should be visible");
    });

    testWidgets('fling on scroll to close bottom sheet', (WidgetTester tester) async {
      app.main();
      await tester.pumpAndSettle();
      final fab = await _openBottomSheet(tester);

      final scrollableZone = find.byKey(const Key("ScrollBox"));
      expect(fab.hitTestable(), findsNothing, reason: "FAB should be invisible");
      expect(scrollableZone, findsOneWidget, reason: "Bottom sheet should be opened");

      await tester.timedDrag(scrollableZone, const Offset(0, 200), const Duration(milliseconds: 500));
      await tester.pumpAndSettle();
      expect(scrollableZone.hitTestable(), findsOneWidget);

      await tester.timedDrag(scrollableZone, const Offset(0, 200), const Duration(milliseconds: 100));
      await tester.pumpAndSettle();

      expect(fab.hitTestable(), findsOneWidget, reason: "FAB should be visible");
    });
  });
}

Future<Finder> _openBottomSheet(WidgetTester tester) async {
  final Finder fab = find.byTooltip('Increment');

  await tester.tap(fab);
  await tester.pumpAndSettle();

  return fab;
}


test/just_bottom_sheet_test.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:just_bottom_sheet/drag_zone_position.dart';

import 'package:just_bottom_sheet/just_bottom_sheet.dart';
import 'package:just_bottom_sheet/just_bottom_sheet_configuration.dart';

void main() {
  const draggableZoneKey = Key("DraggableZone");
  const scrollableZoneKey = Key("ScrollZone");
  group(
    'JustBottomSheet tests',
    () {
      testWidgets(
        "bottom sheet opens up",
        (WidgetTester tester) async {
          await tester.pumpWidget(const _ExampleApp());
          await tester.pumpAndSettle();
          final fab = await _openBottomSheet(tester);

          final draggableZone = find.byKey(draggableZoneKey);
          expect(fab.hitTestable(), findsNothing,
              reason: "FAB should be invisible");
          expect(draggableZone, findsOneWidget,
              reason: "Bottom sheet should be opened");
        },
      );
      testWidgets(
        'fling on draggable zone inside to close bottom sheet',
        (WidgetTester tester) async {
          await tester.pumpWidget(const _ExampleApp());
          await tester.pumpAndSettle();
          final fab = await _openBottomSheet(tester);

          final draggableZone = find.byKey(draggableZoneKey);
          expect(fab.hitTestable(), findsNothing,
              reason: "FAB should be invisible");
          expect(draggableZone, findsOneWidget,
              reason: "Bottom sheet should be opened");

          await tester.fling(draggableZone, const Offset(0, 50), 500);
          await tester.pumpAndSettle();
          expect(draggableZone.hitTestable(), findsOneWidget);

          await tester.fling(draggableZone, const Offset(0, 50), 1500);
          await tester.pumpAndSettle();

          expect(fab.hitTestable(), findsOneWidget,
              reason: "FAB should be visible");
        },
      );
      testWidgets(
        'fling on draggable zone outside to close bottom sheet',
        (WidgetTester tester) async {
          await tester.pumpWidget(const _ExampleApp(
            dragZonePosition: DragZonePosition.outside,
          ));
          await tester.pumpAndSettle();
          final fab = await _openBottomSheet(tester);

          final draggableZone = find.byKey(draggableZoneKey);
          expect(fab.hitTestable(), findsNothing,
              reason: "FAB should be invisible");
          expect(draggableZone, findsOneWidget,
              reason: "Bottom sheet should be opened");

          await tester.fling(draggableZone, const Offset(0, 50), 500);
          await tester.pumpAndSettle();
          expect(draggableZone.hitTestable(), findsOneWidget);

          await tester.fling(draggableZone, const Offset(0, 50), 1500);
          await tester.pumpAndSettle();

          expect(fab.hitTestable(), findsOneWidget,
              reason: "FAB should be visible");
        },
      );
      testWidgets(
        'drag on scroll to close bottom sheet',
        (WidgetTester tester) async {
          await tester.pumpWidget(const _ExampleApp());
          await tester.pumpAndSettle();
          final fab = await _openBottomSheet(tester);

          final scrollableZone = find.byKey(scrollableZoneKey);
          expect(fab.hitTestable(), findsNothing,
              reason: "FAB should be invisible");
          expect(scrollableZone, findsOneWidget,
              reason: "Bottom sheet should be opened");

          await tester.timedDrag(
            scrollableZone,
            const Offset(0, 200),
            const Duration(milliseconds: 500),
          );
          await tester.pumpAndSettle();
          expect(scrollableZone.hitTestable(), findsOneWidget);

          await tester.timedDrag(
            scrollableZone,
            const Offset(0, 200),
            const Duration(milliseconds: 80),
          );
          await tester.pumpAndSettle();

          expect(
            fab.hitTestable(),
            findsOneWidget,
            reason: "FAB should be visible",
          );
        },
      );
      testWidgets(
        'drag on graggable zone to pull up',
        (WidgetTester tester) async {
          await tester.pumpWidget(const _ExampleApp());
          await tester.pumpAndSettle();
          final fab = await _openBottomSheet(tester);

          final dragZone = find.byKey(draggableZoneKey);
          expect(fab.hitTestable(), findsNothing,
              reason: "FAB should be invisible");
          expect(dragZone, findsOneWidget,
              reason: "Bottom sheet should be opened");

          await tester.timedDrag(
            dragZone,
            const Offset(0, -1000),
            const Duration(milliseconds: 100),
          );
          expect(dragZone.hitTestable(), findsOneWidget);
          await tester.pumpAndSettle();
          expect(dragZone.hitTestable(), findsOneWidget);
        },
      );
      testWidgets(
        'drag on scroll up and down repeatedly - should not be closed',
        (WidgetTester tester) async {
          await tester.pumpWidget(const _ExampleApp());
          await tester.pumpAndSettle();
          await _openBottomSheet(tester);

          final scrollableZone = find.byKey(scrollableZoneKey);
          expect(scrollableZone, findsOneWidget,
              reason: "Bottom sheet should be opened");

          await tester.timedDrag(
            scrollableZone,
            const Offset(0, 100),
            const Duration(milliseconds: 200),
          );
          await tester.timedDrag(
            scrollableZone,
            const Offset(0, -200),
            const Duration(milliseconds: 50),
          );
          await tester.timedDrag(
            scrollableZone,
            const Offset(0, 200),
            const Duration(milliseconds: 200),
          );
          await tester.timedDrag(
            scrollableZone,
            const Offset(0, -200),
            const Duration(milliseconds: 50),
          );
          await tester.timedDrag(
            scrollableZone,
            const Offset(0, 200),
            const Duration(milliseconds: 200),
          );
          await tester.timedDrag(
            scrollableZone,
            const Offset(0, -200),
            const Duration(milliseconds: 50),
          );
          await tester.pumpAndSettle();

          expect(scrollableZone, findsOneWidget,
              reason: "Bottom sheet should be opened");
        },
      );
      testWidgets(
        'drag on scroll to close bottom sheet, but it should not close',
        (WidgetTester tester) async {
          await tester.pumpWidget(const _ExampleApp(
            closeOnScroll: false,
          ));
          await tester.pumpAndSettle();
          final fab = await _openBottomSheet(tester);

          final scrollableZone = find.byKey(scrollableZoneKey);
          expect(fab.hitTestable(), findsNothing,
              reason: "FAB should be invisible");
          expect(scrollableZone, findsOneWidget,
              reason: "Bottom sheet should be opened");

          await tester.timedDrag(
            scrollableZone,
            const Offset(0, 200),
            const Duration(milliseconds: 500),
          );
          await tester.pumpAndSettle();
          expect(scrollableZone.hitTestable(), findsOneWidget);

          await tester.timedDrag(
            scrollableZone,
            const Offset(0, 200),
            const Duration(milliseconds: 80),
          );
          await tester.pumpAndSettle();

          expect(
            scrollableZone.hitTestable(),
            findsOneWidget,
            reason: "Scroll should be visible",
          );
        },
      );
      testWidgets('using blur causing transparent background for drag zone',
          (WidgetTester tester) async {
        await tester.pumpWidget(const _ExampleApp(
          closeOnScroll: false,
          useBlur: true,
        ));
        await tester.pumpAndSettle();
        await _openBottomSheet(tester);
        final draggableZoneBackground = find.byKey(
          const Key("DraggableZoneBackground"),
        );
        final draggableZoneWidget =
            tester.firstWidget(draggableZoneBackground) as DecoratedBox;
        final decoration = draggableZoneWidget.decoration as BoxDecoration;

        expect(decoration.color, Colors.transparent);
      });
      testWidgets('using blur causing transparent background for drag zone',
          (WidgetTester tester) async {
        await tester.pumpWidget(const _ExampleApp(
          closeOnScroll: false,
          useBlur: true,
        ));
        await tester.pumpAndSettle();
        await _openBottomSheet(tester);
        final draggableZoneBackground = find.byKey(
          const Key("DraggableZoneBackground"),
        );
        final draggableZoneWidget =
            tester.firstWidget(draggableZoneBackground) as DecoratedBox;
        final decoration = draggableZoneWidget.decoration as BoxDecoration;

        expect(decoration.color, Colors.transparent);
      });
      testWidgets('not using blur causing white background for drag zone',
          (WidgetTester tester) async {
        await tester.pumpWidget(const _ExampleApp(
          closeOnScroll: false,
          useBlur: false,
        ));
        await tester.pumpAndSettle();
        await _openBottomSheet(tester);
        final draggableZoneBackground = find.byKey(
          const Key("DraggableZoneBackground"),
        );
        final draggableZoneWidget =
            tester.firstWidget(draggableZoneBackground) as DecoratedBox;
        final decoration = draggableZoneWidget.decoration as BoxDecoration;

        expect(decoration.color, Colors.white, reason: "Color should be white");
      });
    },
  );
}

Future<Finder> _openBottomSheet(WidgetTester tester) async {
  final Finder fab = find.byTooltip('Increment');

  await tester.tap(fab);
  await tester.pumpAndSettle();

  return fab;
}

class _ExampleApp extends StatefulWidget {
  const _ExampleApp({
    this.closeOnScroll = true,
    this.useBlur = false,
    this.dragZonePosition = DragZonePosition.inside,
    Key? key,
  }) : super(key: key);

  final bool closeOnScroll;
  final bool useBlur;
  final DragZonePosition dragZonePosition;

  @override
  State<_ExampleApp> createState() => _ExampleAppState();
}

class _ExampleAppState extends State<_ExampleApp> {
  final ScrollController scrollController = ScrollController();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(brightness: Brightness.light),
      home: Builder(builder: (context) {
        return Scaffold(
          appBar: AppBar(
            title: const Text("JustBottomSheet Test"),
          ),
          floatingActionButton: FloatingActionButton(
            onPressed: () {
              showJustBottomSheet(
                  context: context,
                  dragZoneConfiguration: JustBottomSheetDragZoneConfiguration(
                    dragZonePosition: widget.dragZonePosition,
                  ),
                  configuration: JustBottomSheetPageConfiguration(
                    closeOnScroll: widget.closeOnScroll,
                    cornerRadius: 32,
                    scrollController: scrollController,
                    height: MediaQuery.of(context).size.height,
                    backgroundImageFilter: widget.useBlur
                        ? ImageFilter.blur(
                            sigmaX: 0,
                            sigmaY: 0,
                          )
                        : null,
                    backgroundColor: Colors.white,
                    builder: (context) {
                      return ListView.builder(
                        controller: scrollController,
                        itemBuilder: (context, row) {
                          return Material(
                            color: Colors.transparent,
                            child: ListTile(
                              title: Text("Row #$row"),
                            ),
                          );
                        },
                        itemCount: 99,
                      );
                    },
                  ));
            },
            tooltip: 'Increment',
            child: const Icon(Icons.add),
          ),
        );
      }),
    );
  }
}


example/lib/main.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:just_bottom_sheet/drag_zone_position.dart';
import 'package:just_bottom_sheet/just_bottom_sheet.dart';
import 'package:just_bottom_sheet/just_bottom_sheet_configuration.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'JustBottomSheet Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        backgroundColor: Colors.white,
      ),
      darkTheme: ThemeData(
        brightness: Brightness.dark,
      ),
      home: const MyHomePage(title: 'JustBottomSheet Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final scrollController = ScrollController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Image.asset("assets/images/flutter.png"),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          showJustBottomSheet(
            context: context,
            dragZoneConfiguration: JustBottomSheetDragZoneConfiguration(
              dragZonePosition: DragZonePosition.outside,
              child: ClipRRect(
                borderRadius: BorderRadius.circular(4),
                child: Container(
                  height: 4,
                  width: 30,
                  color: Theme.of(context).brightness == Brightness.light
                      ? Colors.grey[300]
                      : Colors.white,
                ),
              ),
            ),
            configuration: JustBottomSheetPageConfiguration(
              height: MediaQuery.of(context).size.height,
              builder: (context) {
                return ListView.builder(
                  padding: EdgeInsets.zero,
                  controller: scrollController,
                  itemBuilder: (context, row) {
                    if (row == 5) {
                      return const _NestedPageView();
                    }
                    return Material(
                      color: Colors.transparent,
                      child: ListTile(
                        title: Text("Row #$row"),
                      ),
                    );
                  },
                  itemCount: 25,
                );
              },
              scrollController: scrollController,
              closeOnScroll: true,
              cornerRadius: 16,
              backgroundColor: Theme.of(context).canvasColor.withOpacity(0.5),
              backgroundImageFilter: ImageFilter.blur(
                sigmaX: 30,
                sigmaY: 30,
              ),
            ),
          );
        },
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

class _NestedPageView extends StatefulWidget {
  const _NestedPageView({Key? key}) : super(key: key);

  @override
  State<_NestedPageView> createState() => _NestedPageViewState();
}

class _NestedPageViewState extends State<_NestedPageView> {
  @override
  Widget build(BuildContext context) {
    return Material(
      child: SizedBox(
        height: 300,
        child: PageView(
          children: [
            Container(
              width: 300,
              color: Colors.green,
              child: const Center(
                child: Text("green"),
              ),
            ),
            Container(
              width: 300,
              color: Colors.red,
              child: const Center(
                child: Text("red"),
              ),
            ),
            Container(
              width: 300,
              color: Colors.orange,
              child: const Center(
                child: Text("orange"),
              ),
            ),
            Container(
              width: 300,
              color: Colors.amber,
              child: const Center(
                child: Text("amber"),
              ),
            ),
            Container(
              width: 300,
              color: Colors.cyan,
              child: const Center(
                child: Text("cyan"),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/just_bottom_sheet_route.dart
import 'package:flutter/material.dart';

import 'just_bottom_sheet.dart';
import 'just_bottom_sheet_configuration.dart';

class JustBottomSheetRoute<T> extends ModalRoute<T> {
  JustBottomSheetRoute({
    required this.configuration,
    required this.dragZoneConfiguration,
  }) : super();

  final JustBottomSheetPageConfiguration configuration;
  final JustBottomSheetDragZoneConfiguration dragZoneConfiguration;

  @override
  Color? get barrierColor => Colors.black.withOpacity(0.5);

  @override
  bool get barrierDismissible => true;

  @override
  String? get barrierLabel => null;

  @override
  Widget buildPage(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation) {
    return JustBottomSheetPage(
      configuration: configuration,
      dragZoneConfiguration: dragZoneConfiguration,
    );
  }

  @override
  Widget buildTransitions(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    var begin = const Offset(0, 1);
    var end = Offset.zero;
    var tween =
        Tween(begin: begin, end: end).chain(CurveTween(curve: Curves.easeOut));
    var offsetAnimation = animation.drive(tween);

    return SlideTransition(
      position: offsetAnimation,
      child: child,
    );
  }

  @override
  bool get maintainState => false;

  @override
  bool get opaque => false;

  @override
  Duration get transitionDuration => const Duration(milliseconds: 250);
}


lib/just_bottom_sheet.dart
library just_bottom_sheet;

import 'dart:math';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:just_bottom_sheet/drag_zone_position.dart';
import 'package:just_bottom_sheet/just_bottom_sheet_drag_zone.dart';

import 'just_bottom_sheet_configuration.dart';
import 'just_bottom_sheet_route.dart';

class JustBottomSheetPage extends StatefulWidget {
  const JustBottomSheetPage({
    required this.configuration,
    required this.dragZoneConfiguration,
    Key? key,
  }) : super(key: key);

  final JustBottomSheetPageConfiguration configuration;
  final JustBottomSheetDragZoneConfiguration dragZoneConfiguration;

  @override
  _JustBottomSheetPageState createState() => _JustBottomSheetPageState();
}

class _JustBottomSheetPageState extends State<JustBottomSheetPage>
    with TickerProviderStateMixin {
  bool isDragging = false;
  bool isClampingScroll = false;
  bool shouldPop = false;
  bool willPop = false;

  bool isPointerDown = false;

  double targetYOffset = 0;
  double previousYOffset = 0;
  double scrollPosition = 0;

  final double velocityToClose = 1500;

  void handleDragUpdates({required double offset, required double delta}) {
    if (offset <= -20) {
      previousYOffset = -20;
      return;
    }

    if ((offset > targetYOffset && delta > 0)) {
      shouldPop = true;
    } else {
      shouldPop = false;
    }
    setState(() {
      previousYOffset = targetYOffset;
      targetYOffset = offset;
    });
  }

  bool isVelocityEnoughToPop(double velocity) {
    return velocity >= velocityToClose;
  }

  void onDragZoneDragStart(DragStartDetails details) {
    targetYOffset = 0;
    previousYOffset = 0;
    isDragging = true;
  }

  void onDragZoneDragEnd(DragEndDetails details) {
    isDragging = false;
    bool shouldPop = this.shouldPop;

    final velocity = details.primaryVelocity ?? 0;

    shouldPop = shouldPop && isVelocityEnoughToPop(velocity) ||
        targetYOffset >= widget.configuration.height / 2;

    if (shouldPop && !willPop) {
      willPop = true;
      Navigator.of(context).pop();
      return;
    } else {
      setState(() {
        targetYOffset = 0;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    final mediaQuery = MediaQuery.of(context);
    final height = mediaQuery.size.height;
    final offsetCompensation = height - widget.configuration.height;
    final width = mediaQuery.size.width;
    final topSafeAreaPadding = mediaQuery.padding.top;
    final additionalTopPadding = widget.configuration.additionalTopPadding;
    final totalTopPadding = topSafeAreaPadding + additionalTopPadding;

    const double handlerContainerHeight = 32;
    const handlerPadding = EdgeInsets.all(8);
    const double handlerHeight = 8;

    Widget backgroundChild = Builder(
      builder: (context) {
        return Container(
          color: Theme.of(context).canvasColor,
          child: widget.dragZoneConfiguration.dragZonePosition ==
                  DragZonePosition.inside
              ? JustBottomSheetDragZone(
                  dragZoneConfiguration: widget.dragZoneConfiguration.copyWith(
                    backgroundColor:
                        widget.configuration.backgroundImageFilter == null
                            ? widget.dragZoneConfiguration.backgroundColor
                            : Colors.transparent,
                  ),
                  onDragStart: onDragZoneDragStart,
                  onDragUpdate: (details) {
                    final newYOffset = details.globalPosition.dy -
                        totalTopPadding -
                        offsetCompensation -
                        handlerPadding.top -
                        handlerHeight / 2;

                    handleDragUpdates(
                      offset: newYOffset,
                      delta: details.delta.dy,
                    );
                  },
                  onDragEnd: onDragZoneDragEnd,
                  child: widget.dragZoneConfiguration.child,
                )
              : null,
        );
      },
    );

    final backgroundImageFilter = widget.configuration.backgroundImageFilter;
    if (backgroundImageFilter != null) {
      backgroundChild = BackdropFilter(
        filter: backgroundImageFilter,
        child: backgroundChild,
      );
    }

    final Widget backgroundBody = Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        if (widget.dragZoneConfiguration.dragZonePosition ==
            DragZonePosition.outside)
          JustBottomSheetDragZone(
            dragZoneConfiguration: widget.dragZoneConfiguration.copyWith(
              backgroundColor: Colors.transparent,
            ),
            onDragStart: onDragZoneDragStart,
            onDragUpdate: (details) {
              final newYOffset = details.globalPosition.dy -
                  totalTopPadding -
                  offsetCompensation -
                  handlerPadding.top -
                  handlerHeight / 2;

              handleDragUpdates(
                offset: newYOffset,
                delta: details.delta.dy,
              );
            },
            onDragEnd: onDragZoneDragEnd,
            child: widget.dragZoneConfiguration.child,
          ),
        Expanded(
          child: ClipRRect(
            borderRadius: BorderRadius.only(
              topLeft: Radius.circular(widget.configuration.cornerRadius ?? 16),
              topRight:
                  Radius.circular(widget.configuration.cornerRadius ?? 16),
            ),
            child: backgroundChild,
          ),
        ),
      ],
    );

    return Theme(
      data: Theme.of(context).copyWith(
        canvasColor: widget.configuration.backgroundColor,
      ),
      child: Stack(
        children: [
          TweenAnimationBuilder<double>(
            duration: isDragging || scrollPosition < 0
                ? Duration.zero
                : const Duration(milliseconds: 250),
            curve: Curves.linearToEaseOut,
            tween: Tween<double>(
              begin: previousYOffset,
              end: targetYOffset,
            ),
            builder: (context, value, child) {
              double height = widget.configuration.height - totalTopPadding;

              if (targetYOffset <= 0) {
                height = height + value * -1;
              } else {
                height = height - value;
              }

              return Positioned(
                bottom: 0,
                child: SizedBox(
                  width: width,
                  height: max(0, height),
                  child: child!,
                ),
              );
            },
            child: backgroundBody,
          ),
          TweenAnimationBuilder<double>(
            duration:
                isDragging ? Duration.zero : const Duration(milliseconds: 250),
            curve: Curves.linearToEaseOut,
            tween: Tween<double>(
              begin: widget.configuration.height -
                  totalTopPadding -
                  handlerContainerHeight -
                  (scrollPosition < 0 ? 0 : previousYOffset),
              end: widget.configuration.height -
                  totalTopPadding -
                  handlerContainerHeight -
                  (scrollPosition < 0 ? 0 : targetYOffset),
            ),
            builder: (context, value, child) {
              return Positioned(
                bottom: 0,
                child: SizedBox(
                  key: const Key("ScrollZone"),
                  width: width,
                  height: max(0, value),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: [
                      Expanded(
                        child: Listener(
                          onPointerDown: (_) {
                            isPointerDown = true;
                          },
                          onPointerUp: (_) {
                            isPointerDown = false;
                            if (willPop) {
                              Navigator.of(context).pop();
                            }
                          },
                          child: NotificationListener<ScrollUpdateNotification>(
                            onNotification: (details) {
                              if (details.metrics.axisDirection ==
                                      AxisDirection.left ||
                                  details.metrics.axisDirection ==
                                      AxisDirection.right) {
                                return false;
                              }
                              if (!widget.configuration.closeOnScroll) {
                                return false;
                              }
                              final delta =
                                  details.dragDetails?.primaryDelta?.abs() ?? 0;
                              if (details.metrics.pixels <= 0) {
                                if (delta > 30 && !willPop && isPointerDown) {
                                  willPop = true;
                                }
                                scrollPosition = details.metrics.pixels;
                                handleDragUpdates(
                                  offset: scrollPosition * -1,
                                  delta: -1,
                                );
                              } else {
                                willPop = false;
                                if (targetYOffset != 0) {
                                  handleDragUpdates(
                                    offset: 0,
                                    delta: 1,
                                  );
                                }
                                scrollPosition = details.metrics.pixels;
                              }
                              return false;
                            },
                            child: ScrollConfiguration(
                              behavior: const ScrollBehavior().copyWith(
                                physics: BouncingScrollPhysics(
                                  parent: isDragging
                                      ? const NeverScrollableScrollPhysics()
                                      : const AlwaysScrollableScrollPhysics(),
                                ),
                              ),
                              child: widget.configuration.builder(context),
                            ),
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              );
            },
            child: backgroundBody,
          ),
        ],
      ),
    );
  }
}

Future<T?> showJustBottomSheet<T>({
  required BuildContext context,
  required JustBottomSheetPageConfiguration configuration,
  JustBottomSheetDragZoneConfiguration dragZoneConfiguration =
      const JustBottomSheetDragZoneConfiguration(),
}) {
  return Navigator.of(context).push<T>(
    JustBottomSheetRoute(
        configuration: configuration,
        dragZoneConfiguration: dragZoneConfiguration),
  );
}
